Source/DafnyLanguageServer/Handlers/DafnyCodeActionHandler.cs

```diff
@@ -0,0 +1,213 @@
+        Data = new JArray(documentUri, fixWithId.Id),
+        Diagnostics = fixWithId.DafnyCodeAction.Diagnostics
+      };
+      return t;
+    }
+    ).ToArray();
+    return new CommandOrCodeActionContainer(codeActions);
+  }
+
+  private DafnyCodeActionProvider[] GetDafnyCodeActionProviders() {
+    return new List<DafnyCodeActionProvider>() {
+      new VerificationDafnyCodeActionProvider()
+    }.Concat(
+      DafnyOptions.O.Plugins.SelectMany(plugin =>
+        plugin is ConfiguredPlugin { Configuration: PluginConfiguration configuration } ?
+            configuration.GetDafnyCodeActionProviders() : new DafnyCodeActionProvider[] { })).ToArray();
+  }
+
+  public override Task<CodeAction> Handle(CodeAction request, CancellationToken cancellationToken) {
+    var command = request.Data;
+    if (command == null || command.Count() < 2 || command[1] == null) {
+      return Task.FromResult(request);
+    }
+
+    string? documentUri = command[0]?.Value<string>();
+    if (documentUri == null || !documentUriToDafnyCodeActiones.TryGetValue(documentUri, out var quickFixes)) {
+      return Task.FromResult(request);
+    }
+
+    int? id = command[1]?.Value<int>();
+    if (id == null) {
+      return Task.FromResult(request);
+    }
+
+    DafnyCodeActionWithId? selectedDafnyCodeAction = quickFixes.Where(pluginDafnyCodeAction => pluginDafnyCodeAction.Id == id)
+      .FirstOrDefault((DafnyCodeActionWithId?)null);
+    if (selectedDafnyCodeAction == null) {
+      return Task.FromResult(request);
+    }
+
+    return Task.FromResult(new CodeAction {
+      Edit = new WorkspaceEdit() {
+        DocumentChanges = new Container<WorkspaceEditDocumentChange>(
+          new WorkspaceEditDocumentChange(new TextDocumentEdit() {
+            TextDocument = new OptionalVersionedTextDocumentIdentifier() {
+              Uri = documentUri
+            },
+            Edits = new TextEditContainer(GetTextEdits(selectedDafnyCodeAction.DafnyCodeAction.GetEdits()))
+          }))
+      }
+    });
+  }
+
+  private IEnumerable<TextEdit> GetTextEdits(IEnumerable<DafnyCodeActionEdit> quickFixEdits) {
+    var edits = new List<TextEdit>();
+    foreach (var (range, toReplace) in quickFixEdits) {
+      edits.Add(new TextEdit() {
+        NewText = toReplace,
+        Range = range
+      });
+    }
+    return edits;
+  }
+}
+
+public class DafnyCodeActionInput : IDafnyCodeActionInput {
+  public DafnyCodeActionInput(
+    DafnyDocument document) {
+    Document = document;
+  }
+
+  public string Uri => Document.Uri.ToString();
+  public int Version => Document.Version;
+  public string Code => Document.TextDocumentItem.Text;
+  public Dafny.Program Program => Document.Program;
+  public DafnyDocument Document { get; }
+
+  public Diagnostic[] Diagnostics {
+    get {
+      var result = Document.Diagnostics.ToArray();
+      if (result.Length == 0) {
+        // For anonymous documents opened in VSCode
+        result = Document.Diagnostics.ToArray();
```

Comment:
I'm not sure I understand what's happening here. Does this do anything different from line 150?

Reply:
Ahah interesting ! the code there used to be

```cs
      var result = Document.Errors.GetDiagnostics(Document.Uri).ToArray();
      if (result.Length == 0) {
        // For anonymous documents opened in VSCode
        result = Document.Errors.GetDiagnostics(DocumentUri).ToArray();
      }
```
so that it would handle separatedly the document URI for anonymous documents. However, since the recent refactorings, and the fact that URI are universally used, the case split is no longer user. Thanks for pointing out.